src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -178,11 +181,20 @@ private static void checkpointRestore1() throws
     public static void checkpointRestore() throws
             CheckpointException,
             RestoreException {
-        try {
-            checkpointRestore1();
-        } finally {
-            if (traceStartupTime) {
-                System.out.println("STARTUPTIME " + System.nanoTime() + " restore-finish");
+        synchronized (checkpointRestoreLock) {
+            if (!checkpointInProgress) {
+                try {
+                    checkpointInProgress = true;
+                    checkpointRestore1();
+                } finally {
+                    if (traceStartupTime) {
+                        System.out.println("STARTUPTIME " + System.nanoTime() + " restore-finish");
+                    }
+                    checkpointInProgress = false;
+                }
+            } else {
+                // Recursive checkpoint/restore is not allowed
```

Comment:
Not sure if this comment clarifies anything...

Reply:
Sure, removed